spring-cloud-zookeeper-discovery/src/main/java/org/springframework/cloud/zookeeper/discovery/configclient/ZookeeperConfigServerBootstrapper.java

```diff
@@ -123,11 +79,19 @@ public void initialize(BootstrapRegistry registry) {
 
 		// promote beans to context
 		registry.addCloseListener(event -> {
-			ZookeeperDiscoveryClient discoveryClient = event.getBootstrapContext().get(ZookeeperDiscoveryClient.class);
+			ZookeeperFunction zookeeperFunction = ((ZookeeperFunction) event.getBootstrapContext().get(ConfigServerInstanceProvider.Function.class));
```

Comment:
I think we can make it public and non-final then.  That would resolve this, agree?

Reply:
It would be much better if we could completely avoid this "mini-context" class somehow (hence my other ideas in the other comments), but yes, I think doing so would help if `ZookeeperFunction` has to stay.